linksystem.go

```diff
@@ -0,0 +1,193 @@
+package ipld
+
+import (
+	"fmt"
+	"hash"
+	"io"
+)
+
+type LinkSystem struct {
+	EncoderChooser     func(LinkPrototype) (Encoder, error)
+	DecoderChooser     func(Link) (Decoder, error)
+	HasherChooser      func(LinkPrototype) (hash.Hash, error)
+	StorageWriteOpener BlockWriteOpener
+	StorageReadOpener  BlockReadOpener
```

Comment:
Review wanted on naming here.  (Both the functions, and the field names.)

@hannahhoward had previously commented on them being a bit of a mouthful, but rather liking that they mention "blocks" now.  I think I've improved the names a bit since the draft she was looking at, but they're still a mouthful.

Probably I should `s/Storage/Block/` on the field names, so they're all just consistent?

Reply:
Somewhat unrelated, and perhaps a bit of a nit: I don't think func fields should be named like interfaces. It's far more natural to call `lsys.ChooseEncoder(...)` or `lsys.EncoderChoose(...)` than `lsys.EncoderChooser(...)`. The result of that call is an encoder, not an encoder chooser :) A bit like if `strings.Repeat` were called `strings.Repeater`.

Interfaces are different, because you call methods on them, and it's good to differentiate the interface name from the method name. You just have one name here, so in my opinion it should be what best fits the method-like function calls.